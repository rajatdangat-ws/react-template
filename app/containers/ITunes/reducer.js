/*
 *
 * ITunes reducer
 *
 */
import produce from 'immer';
import get from 'lodash/get';
import { createActions } from 'reduxsauce';

export const initialState = { songs: [], error: null };

export const { Types: iTunesTypes, Creators: iTunesCreators } = createActions({
  requestGetSongs: ['searchTerm'],
  successGetSongs: ['data'],
  errorGetSongs: ['error'],
  clearSongs: null
});

export const iTunesReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case iTunesTypes.SUCCESS_GET_SONGS:
        draft.songs = action.data;
        break;
      case iTunesTypes.ERROR_GET_SONGS:
        draft.error = get(action.error, 'message', 'something_went_wrong');
        break;
      case iTunesTypes.CLEAR_SONGS:
        draft.searchTerm = null;
        draft.error = null;
        draft.songs = [];
        break;
      default:
        return state;
    }
  });

export default iTunesReducer;
