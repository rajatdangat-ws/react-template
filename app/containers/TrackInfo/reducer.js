/*
 *
 * TrackInfo reducer
 *
 */
import produce from 'immer';
import get from 'lodash/get';
import { createActions } from 'reduxsauce';

export const initialState = { songDetails: {}, error: null };

export const { Types: trackInfoTypes, Creators: trackInfoCreators } = createActions({
  requestGetSongDetails: ['songId', 'songDetails'],
  successGetSongDetails: ['data'],
  errorGetSongDetails: ['error'],
  clearSongDetails: {}
});

/* eslint-disable default-case, no-param-reassign */
export const trackInfoReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case trackInfoTypes.SUCCESS_GET_SONG_DETAILS:
        draft.songDetails = action.data;
        break;
      case trackInfoTypes.ERROR_GET_SONG_DETAILS:
        draft.error = get(action.error, 'message', 'something_went_wrong');
        break;
      case trackInfoTypes.CLEAR_SONG_DETAILS:
        draft.error = null;
        draft.songDetails = {};
        break;
      default:
        return state;
    }
  });

export default trackInfoReducer;
