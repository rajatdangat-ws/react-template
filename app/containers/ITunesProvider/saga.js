import { takeLatest, call, put } from 'redux-saga/effects';
import { getSongs } from '@app/services/itunesApi';
import { iTunesTypes, iTunesCreators } from './reducer';

const { REQUEST_GET_SONGS } = iTunesTypes;
const { successGetSongs, errorGetSongs } = iTunesCreators;

export function* getSongsFunction(action) {
  const response = yield call(getSongs, action.searchTerm);
  const { data, ok } = response;
  if (ok) {
    const results = data?.results || [];
    const newResults = results.map((item) => {
      return {
        ...item,
        artworkUrl250: item.artworkUrl100.replace('100x100bb.jpg', '250x250bb.jpg')
      };
    });
    yield put(successGetSongs(newResults));
  } else {
    yield put(errorGetSongs(data));
  }
}

export default function* iTunesProviderSaga() {
  yield takeLatest(REQUEST_GET_SONGS, getSongsFunction);
}
