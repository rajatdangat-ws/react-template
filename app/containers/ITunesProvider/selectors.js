import { createSelector } from 'reselect';
import get from 'lodash/get';
import { initialState } from './reducer';

/**
 * Direct selector to the iTunesProvider state domain
 */

const selectITunesProviderDomain = (state) => state.iTunesProviderContainer || initialState;

const makeSelectITunesProvider = () => createSelector(selectITunesProviderDomain, (substate) => substate);

export default makeSelectITunesProvider;
export { selectITunesProviderDomain };

export const selectSongs = () => createSelector(selectITunesProviderDomain, (substate) => get(substate, 'songs'));

export const selectSongById = (songId) =>
  createSelector(selectSongs(), (songs) => {
    return songs.find((song) => song.trackId == songId);
  });

export const selectError = () => createSelector(selectITunesProviderDomain, (substate) => get(substate, 'error'));

export const selectSearchTerm = () =>
  createSelector(selectITunesProviderDomain, (substate) => get(substate, 'searchTerm'));
