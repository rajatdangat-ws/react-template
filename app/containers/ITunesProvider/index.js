/**
 *
 * ITunesProvider
 *
 */

import React, { memo } from 'react';
// import PropTypes from 'prop-types'
import { connect } from 'react-redux';
import { injectIntl, FormattedMessage as T } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import { useInjectSaga } from '@utils/injectSaga';
import makeSelectITunesProvider from './selectors';
import saga from './saga';

export function ITunesProvider() {
  useInjectSaga({ key: 'iTunesProvider', saga });

  return (
    <div>
      <T id={'ITunesProvider'} />
    </div>
  );
}

ITunesProvider.propTypes = {};

const mapStateToProps = createStructuredSelector({
  iTunesProvider: makeSelectITunesProvider()
});

function mapDispatchToProps(dispatch) {
  return {
    dispatch
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect, memo)(ITunesProvider);

export const ITunesProviderTest = compose(injectIntl)(ITunesProvider);
