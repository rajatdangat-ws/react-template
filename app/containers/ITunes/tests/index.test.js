/**
 *
 * Tests for ITunes
 *
 *
 */

import React from 'react';
import { renderProvider, timeout } from '@utils/testUtils';
// import { fireEvent } from '@testing-library/dom'
import userEvent from '@testing-library/user-event';
import { ITunesTest as ITunes, mapDispatchToProps } from '../index';
import { iTunesTypes } from '@app/containers/ITunesProvider/reducer';

const songs = [
  {
    trackName: 'test track 1',
    trackId: 1,
    coverImgUrl: 'image url 1',
    artistName: 'test artist 1',
    previewUrl: 'preview url 1',
    detailsUrl: 'details url 1'
  },
  {
    trackName: 'test track 2',
    trackId: 2,
    coverImgUrl: 'image url 2',
    artistName: 'test artist 2',
    previewUrl: 'preview url 2',
    detailsUrl: 'details url 2'
  }
];

describe('<ITunes /> container tests', () => {
  // let submitSpy

  beforeEach(() => {
    // submitSpy = jest.fn()
  });
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<ITunes songs={songs} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should have search field', () => {
    const { getByTestId } = renderProvider(
      <ITunes dispatchRequestGetSongs={() => {}} songs={songs} dispatchClearSongs={() => {}} />
    );
    const searchField = getByTestId('search-field');
    expect(searchField).toBeInTheDocument();
    expect(searchField).toHaveAttribute('type', 'text');
  });

  it('should dispatch get action on search', async () => {
    const getSongsSpy = jest.fn();
    const { getByRole, getByTestId } = renderProvider(
      <ITunes dispatchRequestGetSongs={getSongsSpy} songs={songs} dispatchClearSongs={() => {}} />
    );
    const user = userEvent.setup();
    const searchButton = getByRole('button', { name: 'search' });
    const searchField = getByTestId('search-field');
    user.type(searchField, 'test');

    expect(searchButton).not.toBeNull();
    await user.click(searchButton);

    expect(getSongsSpy).toBeCalled();
  });

  it('should dispatch clear action when search field is empty', async () => {
    const clearSongsSpy = jest.fn();
    const { getByRole } = renderProvider(
      <ITunes dispatchRequestGetSongs={() => {}} songs={songs} dispatchClearSongs={clearSongsSpy} />
    );
    const user = userEvent.setup();
    const searchButton = getByRole('button', { name: 'search' });

    await user.click(searchButton);
    expect(clearSongsSpy).toBeCalled();
  });

  it('should render the grid layout', () => {
    const { getAllByTestId } = renderProvider(
      <ITunes dispatchRequestGetSongs={() => {}} songs={songs} dispatchClearSongs={() => {}} />
    );
    expect(getAllByTestId('music-info-card').length).toEqual(songs.length);
  });

  it('should validate mapDispatchToProps actions', async () => {
    const dispatch = jest.fn();
    const searchTerm = 'linkin park';
    const actions = {
      dispatchRequestGetSongs: { searchTerm, type: iTunesTypes.REQUEST_GET_SONGS },
      dispatchClearSongs: { type: iTunesTypes.CLEAR_SONGS }
    };

    const props = mapDispatchToProps(dispatch);
    props.dispatchRequestGetSongs(searchTerm);
    expect(dispatch).toHaveBeenCalledWith(actions.dispatchRequestGetSongs);

    await timeout(500);
    props.dispatchClearSongs();
    expect(dispatch).toHaveBeenCalledWith(actions.dispatchClearSongs);
  });

  it('should pause the previous playing music when playing new music', async () => {
    const { getAllByTestId } = renderProvider(<ITunes songs={songs} />);
    await timeout(500);
    // const audioElements = getAllByTestId('audio-element');
    // let firstAudioElement = audioElements[0];
    // let secondAudioElement = audioElements[1];
    const user = userEvent.setup();

    const playBtns = getAllByTestId('play-button');
    let firstPlayBtn = playBtns[0];
    let secondPlayBtn = playBtns[1];

    expect(firstPlayBtn).toBeInTheDocument();
    await user.click(firstPlayBtn);
    // expect(firstAudioElement.paused).toBe(false);
    const pauseBtn = getAllByTestId('pause-button')[0];
    expect(pauseBtn).toBeInTheDocument();

    await timeout(500);

    expect(secondPlayBtn).toBeInTheDocument();
    // expect(secondAudioElement.paused).toBe(true);
    await user.click(secondPlayBtn);
    // expect(firstAudioElement.paused).toBe(true);
    // expect(secondAudioElement.paused).toBe(false);

    expect(pauseBtn).not.toBeInTheDocument();
    firstPlayBtn = getAllByTestId('play-button')[0];
    expect(firstPlayBtn).toBeInTheDocument();
    expect(secondPlayBtn).not.toBeInTheDocument();
  });
});
