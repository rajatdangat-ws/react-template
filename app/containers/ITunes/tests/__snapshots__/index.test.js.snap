// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ITunes /> container tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ITunes__SearchContainer-fbwjmv-0 hdgZyn"
    >
      <span
        class="ant-input-group-wrapper ant-input-search ITunes__SearchBox-fbwjmv-2 fpvxAJ"
      >
        <span
          class="ant-input-wrapper ant-input-group"
        >
          <input
            class="ant-input"
            data-testid="search-field"
            placeholder="search for songs..."
            type="text"
            value=""
          />
          <span
            class="ant-input-group-addon"
          >
            <button
              class="ant-btn ant-btn-icon-only ant-input-search-button"
              type="button"
            >
              <span
                aria-label="search"
                class="anticon anticon-search"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="search"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </span>
      </span>
    </div>
    <div
      class="ITunes__TracksListContainer-fbwjmv-1 pftw"
    >
      <div
        class="ITunes__GridLayout-fbwjmv-3 diIYfV"
      >
        <div
          class="ant-card ant-card-bordered MusicInfoCard__InfoCard-sc-4tgp7y-1 hXHvIE"
          data-testid="music-info-card"
        >
          <div
            class="ant-card-cover"
          >
            <img
              alt="test track 1"
              height="250px"
              loading="lazy"
              style="max-inline-size: 100%;"
              width="250px"
            />
          </div>
          <div
            class="ant-card-body"
          >
            <div
              class="ant-card-meta MusicInfoCard__StyledMeta-sc-4tgp7y-0 kOaJye"
            >
              <div
                class="ant-card-meta-detail"
              >
                <div
                  class="ant-card-meta-title"
                >
                  test track 1
                </div>
                <div
                  class="ant-card-meta-description"
                >
                  test artist 1
                </div>
              </div>
            </div>
          </div>
          <ul
            class="ant-card-actions"
          >
            <li
              style="width: 50%;"
            >
              <span>
                <span
                  aria-label="play-circle"
                  class="anticon anticon-play-circle"
                  data-testid="play-button"
                  role="img"
                  tabindex="-1"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="play-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                    <path
                      d="M719.4 499.1l-296.1-215A15.9 15.9 0 00398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9 15.9 0 000-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z"
                    />
                  </svg>
                </span>
              </span>
            </li>
            <li
              style="width: 50%;"
            >
              <span>
                <span
                  aria-label="info-circle"
                  class="anticon anticon-info-circle"
                  data-testid="more-button"
                  role="img"
                  tabindex="-1"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="info-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                    <path
                      d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                    />
                  </svg>
                </span>
              </span>
            </li>
          </ul>
        </div>
        <audio
          data-testid="audio-element"
          src="preview url 1"
        />
        <div
          class="ant-card ant-card-bordered MusicInfoCard__InfoCard-sc-4tgp7y-1 hXHvIE"
          data-testid="music-info-card"
        >
          <div
            class="ant-card-cover"
          >
            <img
              alt="test track 2"
              height="250px"
              loading="lazy"
              style="max-inline-size: 100%;"
              width="250px"
            />
          </div>
          <div
            class="ant-card-body"
          >
            <div
              class="ant-card-meta MusicInfoCard__StyledMeta-sc-4tgp7y-0 kOaJye"
            >
              <div
                class="ant-card-meta-detail"
              >
                <div
                  class="ant-card-meta-title"
                >
                  test track 2
                </div>
                <div
                  class="ant-card-meta-description"
                >
                  test artist 2
                </div>
              </div>
            </div>
          </div>
          <ul
            class="ant-card-actions"
          >
            <li
              style="width: 50%;"
            >
              <span>
                <span
                  aria-label="play-circle"
                  class="anticon anticon-play-circle"
                  data-testid="play-button"
                  role="img"
                  tabindex="-1"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="play-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                    <path
                      d="M719.4 499.1l-296.1-215A15.9 15.9 0 00398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9 15.9 0 000-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z"
                    />
                  </svg>
                </span>
              </span>
            </li>
            <li
              style="width: 50%;"
            >
              <span>
                <span
                  aria-label="info-circle"
                  class="anticon anticon-info-circle"
                  data-testid="more-button"
                  role="img"
                  tabindex="-1"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="info-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                    />
                    <path
                      d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
                    />
                  </svg>
                </span>
              </span>
            </li>
          </ul>
        </div>
        <audio
          data-testid="audio-element"
          src="preview url 2"
        />
      </div>
    </div>
  </div>
</body>
`;
